pkg/resources/process/wait.go

```diff
@@ -69,47 +72,101 @@ func (w *procWaiter) wait(ch chan<- *drivers.ExitResult) {
 	}
 }
 
+const (
+	// exitStatusFile is written into the task directory upon the exit of the
+	// parent task process. It contains the exit status in case the value needs
+	// to be retrieved by the plugin (i.e. if the task process was orphaned).
+	exitStatusFile = ".exit_status.txt"
+)
+
 // WaitPID is able to wait on a given specific PID. We must lookup the
 // process and also send a signal(0) to make sure it is actually still alive
 // before waiting on it.
-func WaitPID(pid int) Waiter {
+func WaitPID(pid int, taskdir string) Waiter {
 	return &pidWaiter{
-		pid: pid,
-		ch:  make(chan *drivers.ExitResult),
+		pid:     pid,
+		taskdir: taskdir,
+		ch:      make(chan *drivers.ExitResult),
 	}
 }
 
 type pidWaiter struct {
-	pid int
-	ch  chan *drivers.ExitResult
+	pid     int
+	taskdir string
+	ch      chan *drivers.ExitResult
 }
 
 func (w *pidWaiter) Wait() WaitCh {
 	go w.wait(w.ch)
 	return w.ch
 }
 
-func (w *pidWaiter) wait(ch chan<- *drivers.ExitResult) {
-	proc, _ := os.FindProcess(w.pid) // never errors on unix
+func unrecoverable(msg string, err error) *drivers.ExitResult {
+	return &drivers.ExitResult{
+		ExitCode: -1,
+		Err:      fmt.Errorf("%s: %w", msg, err),
+	}
+}
 
+func (w *pidWaiter) wait(ch chan<- *drivers.ExitResult) {
 	// send a 0 signal to detect if the process is still alive
+	proc, _ := os.FindProcess(w.pid) // never errors on unix
 	if err := proc.Signal(syscall.Signal(0)); err != nil {
-		// TODO(shoenig): there is still a way to get the exit code and signal
-		// of a recently deceased process by reading /proc; we should look into
-		// that after Nomad 1.8 Beta since Client restarts are a thing that we
-		// support.
-		//
-		// See what the pledge driver is doing for inspiration in
-		// https://github.com/shoenig/nomad-pledge-driver/blob/v0.3.0/pkg/resources/process/wait.go#L82-L105
-		// and make sure it works correctly.
-		ch <- &drivers.ExitResult{
-			ExitCode: -1,
-			Err:      fmt.Errorf("task is gone: %w", err),
-		}
+		ch <- unrecoverable("task process handle is gone", err)
+		return
+	}
+
+	pidfd, err := openProcessFD(w.pid)
+	if err != nil {
+		ch <- unrecoverable("task process file descriptor cannot be opened", err)
+		return
+	}
+
+	pollFD := []unix.PollFd{{
+		Fd:     pidfd,
+		Events: unix.POLLIN,
+	}}
+	const timeout = -1 // infinite
+
+	// no need to check for an error
+	_, _ = unix.Poll(pollFD, timeout)
+
+	// retrieve code from file
+	source := filepath.Join(w.taskdir, "local", exitStatusFile)
```

Comment:
Don't you need to check for this *before* using the pidfd in case a *new* process has started with the same pid as the original task process?

Reply:
Yeah, I think we need to:
* grab the pidfd
* check the exit file is empty while we're holding the pidfd
* then poll the pidfd
* then read the file value when we're done

Also, just double-checking but this PID is the PID of the shim process, right? So we're guaranteed that if the pidfd unblocks that the exit status file has already been written and there's no race between reading and writing?

Reply:
Great ideas! I added 067f9a5 which rewrites this section to take the already-dead (and potentially reborn) process PID into account. I believe doing so also elegantly resolves https://github.com/hashicorp/nomad-driver-exec2/issues/9.

This PID is of the `unshare` process which is a parent of the `shim`. Since `unshare` is configured to `SIGKILL` its children upon its own death the shim will also be dead (and it shouldn't die until the shim exits anyway). I think in practice, we're fine. 

In the next hackathon I'd like to sit down and write a tool that combines `unshare`, `nsenter`, and the landlock parts of the shim into one tool. Probably in Rust so there's no Go runtime overhead.